userapi/src/main/java/model/D42Model.java

```diff
@@ -0,0 +1,117 @@
+    }
+
+    private static Profile getProfile(String key) {
+        String[] splits = key.split(DELIMITER);
+        String start = splits[4];
+        return new Profile(start, ZonedDateTime.parse(start).plusSeconds(Long.parseLong(splits[5])).toString());
+    }
+
+    private String getLastFromCommonPrefix(String commonPrefix) {
+        String[] splits = commonPrefix.split(DELIMITER);
+        return splits[splits.length - 1];
+    }
+
+    private Set<String> getListingAtLevelWithPrefix(String level, String objPrefix, boolean encoded) throws Exception {
+        Set<String> commonPrefixes = store.listAsync(level, false).get(TIMEOUT, TimeUnit.SECONDS);
+        Set<String> objs = new HashSet<>();
+        for (String commonPrefix : commonPrefixes) {
+            String objName = getLastFromCommonPrefix(commonPrefix);
+            if (encoded) {
+                objName = new String(base32.decode(objName.getBytes()));
+            }
+            if (objName.indexOf(objPrefix) == 0) {
+                objs.add(objName);
+            }
+        }
+        return objs;
+    }
+
+    @Override
+    public void setStorage(AsyncStorage store) {
+        this.store = (S3AsyncStorage) store;
+    }
+
+    @Override
+    public Set<String> getAppIdsWithPrefix(String appIdPrefix) throws Exception {
+        return getListingAtLevelWithPrefix(BUCKET + PATH_SEPARATOR + VERSION + DELIMITER, appIdPrefix, true);
+    }
+
+    @Override
+    public Set<String> getClusterIdsWithPrefix(String appId, String clusterIdPrefix) throws Exception {
+        return getListingAtLevelWithPrefix(
+                BUCKET + PATH_SEPARATOR + VERSION + DELIMITER + new String(base32.encode(appId.getBytes())) + DELIMITER,
+                clusterIdPrefix,
+                true);
+    }
+
+    @Override
+    public Set<String> getProcsWithPrefix(String appId, String clusterId, String procPrefix) throws Exception {
+        return getListingAtLevelWithPrefix(BUCKET + PATH_SEPARATOR + VERSION + DELIMITER + new String(base32.encode(appId.getBytes()))
+                        + DELIMITER + new String(base32.encode(clusterId.getBytes())) + DELIMITER,
+                procPrefix,
+                false);
+    }
+
+    @Override
+    public Set<Profile> getProfilesInTimeWindow(String appId, String clusterId, String proc, String startTime, String durationInSeconds) throws Exception {
+        String prefix = BUCKET + PATH_SEPARATOR + VERSION + DELIMITER + new String(base32.encode(appId.getBytes()))
+                + DELIMITER + new String(base32.encode(clusterId.getBytes())) + DELIMITER + proc + DELIMITER;
+        Map<Profile, Set<String>> profilesMap = store.listAsync(prefix, true).get().stream().collect(Collectors.groupingBy(D42Model::getProfile, Collectors.mapping(D42Model::getWorkType, Collectors.toSet())));
+        Set<Profile> profiles = profilesMap.entrySet().stream().map(profileSetEntry -> {
+            profileSetEntry.getKey().setValues(profileSetEntry.getValue());
```

Comment:
I have made the grouping as one-liner to avoid misunderstanding in readability due to second line. Since the groupingby (which is necessary) returns a map, it is re-streamed to make the keys collect the values inside the key object as their member field.